salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+"""
+Rabbitmq transport classes
+This is a copy/modify of zeromq implementation with zeromq-specific bits removed.
+TODO: refactor transport implementations so that tcp, zeromq, rabbitmq share common code
+"""
+import copy
+import errno
+import hashlib
+import logging
+import os
+import signal
+import sys
+import threading
+import weakref
+from random import randint
+from uuid import uuid4
+
+import pika
+from pika import SelectConnection, BasicProperties
+from pika.exchange_type import ExchangeType
+
+import salt.auth
+import salt.crypt
+import salt.ext.tornado
+import salt.ext.tornado.concurrent
+import salt.ext.tornado.gen
+import salt.ext.tornado.ioloop
+import salt.log.setup
+import salt.payload
+import salt.transport.client
+import salt.transport.mixins.auth
+import salt.transport.server
+import salt.utils.event
+import salt.utils.files
+import salt.utils.minions
+import salt.utils.process
+import salt.utils.stringutils
+import salt.utils.verify
+import salt.utils.versions
+import salt.utils.zeromq
+from salt.exceptions import SaltException, SaltReqTimeoutError
+from salt.ext import tornado
+
+try:
+    from M2Crypto import RSA
+
+    HAS_M2 = True
+except ImportError:
+    HAS_M2 = False
+    try:
+        from Cryptodome.Cipher import PKCS1_OAEP
+    except ImportError:
+        from Crypto.Cipher import PKCS1_OAEP  # nosec
+
+log = logging.getLogger(__name__)
+
+
+def _get_master_uri(master_ip, master_port):
+    """
+    TODO: this method is a zeromq remnant and likely should be removed
+    """
+    from salt.utils.zeromq import ip_bracket
+
+    master_uri = "tcp://{master_ip}:{master_port}".format(
+        master_ip=ip_bracket(master_ip), master_port=master_port
+    )
+
+    return master_uri
+
+
+class RMQConnectionWrapperBase:
+    def __init__(self, opts):
+        self._opts = opts
+
+        # rmq broker address
+        self._host = self._opts["transport_rabbitmq_address"] if "transport_rabbitmq_address" \
+                                                                 in self._opts else "localhost"
+        assert self._host
+
+        # rmq broker credentials (TODO: support other types of auth. eventually)
+        assert "transport_rabbitmq_auth" in self._opts
+        creds = self._opts["transport_rabbitmq_auth"]
+        assert "username" in creds and "password" in creds
+        self._creds = pika.PlainCredentials(creds["username"], creds["password"])
+
+
+        # TODO: figure out a better way to generate common queue names that are accessible to both master and minion,
+        # each possibly running on different machine
+        master_port = self._opts["master_port"] if "master_port" in self._opts else self._opts["ret_port"]
+        assert master_port
+
+        self._queue_name = "{}_{}".format(self._host, master_port)
+
+    @property
+    def queue_name(self):
+        return self._queue_name
+
+    def __del__(self):
```

Comment:
**suggestion** `__del__` is never guaranteed to be called in Python - if guarantees are desired then `with` blocks are required.